### PR TITLE
fix the searchable fields bug when a field is nested

### PR DESCRIPTION
**Pull request by irevoire**
I made the PR based on release-v0.26.4, but I guess we don't want to merge it on this branch and instead update it to v0.26.5? :thinking:

### CODE REVIEW

File: milli/src/index.rs [hunk 1]
```diff
@@ -457,12 +458,43 @@ impl Index {
...
     /* searchable fields */
 
-    /// Writes the searchable fields, when this list is specified, only these are indexed.
-    pub(crate) fn put_searchable_fields(
+    /// Write the user defined searchable fields and generate the real searchable fields from the specified fields ids map.
+    pub(crate) fn put_all_searchable_fields_from_fields_ids_map(
         &self,
         wtxn: &mut RwTxn,
-        fields: &[&str],
+        user_fields: &[&str],
```
Comment (Kerollmops) [inline milli/src/index.rs:465]: I would prefer that you directly ask for a `Vec<String>` or `Vec<S> where S: AsRef<str>` here as you `.to_vec()` it just in this function (L476).

File: milli/src/index.rs [hunk 2]
```diff
@@ -498,6 +530,38 @@ impl Index {
...
+    ) -> heed::Result<bool> {
+        self.main.delete::<_, Str>(wtxn, main_key::USER_DEFINED_SEARCHABLE_FIELDS_KEY)
+    }
+
+    /// Returns the user defined searchable fields.
+    pub fn user_defined_searchable_fields<'t>(
+        &self,
+        rtxn: &'t RoTxn,
+    ) -> heed::Result<Option<Vec<&'t str>>> {
+        self.main.get::<_, Str, SerdeBincode<Vec<&'t str>>>(
```
Comment (Kerollmops) [inline milli/src/index.rs:559]:
```suggestion
        self.main.get::<_, Str, SerdeBincode<Vec<_>>>(
```

File: milli/src/index.rs [hunk 3]
```diff
@@ -498,6 +530,38 @@ impl Index {
         }
     }
 
+    /// Writes the searchable fields, when this list is specified, only these are indexed.
+    pub(crate) fn put_user_defined_searchable_fields(
+        &self,
+        wtxn: &mut RwTxn,
+        fields: &[&str],
+    ) -> heed::Result<()> {
+        self.main.put::<_, Str, SerdeBincode<&[&str]>>(
```
Comment (Kerollmops) [inline milli/src/index.rs:539]:
```suggestion
        self.main.put::<_, Str, SerdeBincode<_>>(
```

File: milli/src/update/index_documents/mod.rs [hunk 4]
```diff
@@ -157,6 +157,18 @@ where
...
         self.index.put_faceted_fields(self.wtxn, &new_facets)?;
 
+        // in case new fields were introduced we're going to recreate the searchable fields.
+        if let Some(faceted_fields) = self.index.user_defined_searchable_fields(self.wtxn)? {
+            // we can't keep references on the faceted fields while we update the index thus we need to own it.
+            let faceted_fields: Vec<String> =
+                faceted_fields.into_iter().map(str::to_string).collect();
+            self.index.put_all_searchable_fields_from_fields_ids_map(
+                self.wtxn,
+                &faceted_fields.iter().map(String::as_ref).collect::<Vec<_>>(),
```
Comment (Kerollmops) [inline milli/src/update/index_documents/mod.rs:167]: Doing the previous change would simplify this part a bit. There would not be any need to collect two times.

**Review by Kerollmops — CHANGES_REQUESTED** [review]

**Review by Kerollmops — APPROVED** [review]
bors merge

**Review by Kerollmops — APPROVED** [review]
bors merge

**Comment by curquiza** [comment]
Since we did the PR on an old branch, we still have the issue with bors settings, we need to merge by hand, bors is stuck currently... once we are sure the tests pass, we will merge it by hand

bors try

**Pull request merged**